htop segfaults due to invalid content in htoprc

running htop 2.2 segfaults on rhel7.9:
> [Tue Oct  5 16:29:43 2021] htop[66432]: segfault at 0 ip 00007f69638f7d38 sp 00007ffc7628a6a8 error 4 in libc-2.17.so[7f69637b9000+1c3000]

and 3.0.5 on the same machine:
> [Tue Oct  5 16:23:06 2021] htop[61104]: segfault at 0 ip 00007fd311641d38 sp 00007ffc0d628318 error 4 in libc-2.17.so[7fd311503000+1c3000]

lscpu | grep NUMA
> NUMA node(s):          12
> NUMA node0 CPU(s):     0-23,288-311
> NUMA node1 CPU(s):     24-47,312-335
> NUMA node2 CPU(s):     48-71,336-359
> NUMA node3 CPU(s):     72-95,360-383
> NUMA node4 CPU(s):     96-119,384-407
> NUMA node5 CPU(s):     120-143,408-431
> NUMA node6 CPU(s):     144-167,432-455
> NUMA node7 CPU(s):     168-191,456-479
> NUMA node8 CPU(s):     192-215,480-503
> NUMA node9 CPU(s):     216-239,504-527
> NUMA node10 CPU(s):    240-263,528-551
> NUMA node11 CPU(s):    264-287,552-575

htoprc:
```
# Beware! This file is rewritten by htop when settings are changed in the interface.
# The parser is also very primitive, and not human-friendly.
fields=0 48 17 18 38 39 40 2 46 47 49 1
sort_key=46
sort_direction=1
hide_threads=0
hide_kernel_threads=1
hide_userland_threads=0
shadow_other_users=0
show_thread_names=0
highlight_base_name=0
highlight_megabytes=1
highlight_threads=0
tree_view=0
header_margin=1
detailed_cpu_time=0
cpu_count_from_zero=0
update_process_names=0
account_guest_in_cpu_meter=0
color_scheme=0
delay=15
left_meters=LeftCPUs2 Memory Swap
left_meter_modes=1 1 1 1 1 1
right_meters=RightCPUs2 Tasks LoadAverage Uptime
right_meter_modes=1 2 2 2 2
````
re 3.1.0: I have issues building it:
```

  htop 3.1.0

  platform:                  linux
  os-release file:           /etc/os-release
  (Linux) proc directory:    /proc
  (Linux) openvz:            no
  (Linux) vserver:           no
  (Linux) ancient vserver:   no
  (Linux) delay accounting:  no
  (Linux) sensors:           no
  (Linux) capabilities:      no
  unicode:                   yes
  affinity:                  yes
  hwloc:                     no
  debug:                     yes
  static:                    no

Tue Oct 05 23:02:03 user@:~/htop-3.1.0% make
Makefile:1279: .deps/htop.Po: No such file or directory
make: *** No rule to make target `.deps/htop.Po'.  Stop.

```

when I remove htoprc I get only a black screen. with 2.2 and 3.0.5

--- a/Settings.c
+++ b/Settings.c
@@ -28,11 +28,7 @@ void Settings_delete(Settings* this) {
    free(this->filename);
    free(this->fields);
    for (unsigned int i = 0; i < HeaderLayout_getColumns(this->hLayout); i++) {
-      if (this->hColumns[i].names) {
-         for (size_t j = 0; j < this->hColumns[i].len; j++)
-            free(this->hColumns[i].names[j]);
-         free(this->hColumns[i].names);
-      }
+      String_freeArray(this->hColumns[i].names);
       free(this->hColumns[i].modes);
    }
    free(this->hColumns);
@@ -67,14 +63,26 @@ static void Settings_readMeterModes(Settings* this, const char* line, unsigned i
 
 static void Settings_defaultMeters(Settings* this, unsigned int initialCpuCount) {
    int sizes[] = { 3, 3 };
+
    if (initialCpuCount > 4 && initialCpuCount <= 128) {
       sizes[1]++;
    }
-   for (int i = 0; i < 2; i++) {
+
+   // Release any previously allocated memory
+   for (size_t i = 0; i < HeaderLayout_getColumns(this->hLayout); i++) {
+      String_freeArray(this->hColumns[i].names);
+      free(this->hColumns[i].modes);
+   }
+   free(this->hColumns);
+
+   this->hLayout = HF_TWO_50_50;
+   this->hColumns = xCalloc(HeaderLayout_getColumns(this->hLayout), sizeof(MeterColumnSetting));
+   for (size_t i = 0; i < 2; i++) {
       this->hColumns[i].names = xCalloc(sizes[i] + 1, sizeof(char*));
       this->hColumns[i].modes = xCalloc(sizes[i], sizeof(int));
       this->hColumns[i].len = sizes[i];
    }
+
    int r = 0;
 
    if (initialCpuCount > 128) {

--- a/Settings.c
+++ b/Settings.c
@@ -61,6 +61,29 @@ static void Settings_readMeterModes(Settings* this, const char* line, unsigned i
    this->hColumns[column].modes = modes;
 }
 
+static bool Settings_validateMeters(Settings* this) {
+   const size_t colCount = HeaderLayout_getColumns(this->hLayout);
+
+   for (size_t column = 0; column < colCount; column++) {
+      char** names = this->hColumns[column].names;
+      const int* modes = this->hColumns[column].modes;
+      const size_t len = this->hColumns[column].len;
+
+      if (!names || !modes || !len)
+         return false;
+
+      // Check for each mode there is an entry with a non-NULL name
+      for (size_t meterIdx = 0; meterIdx < len; meterIdx++)
+         if (!names[meterIdx])
+            return false;
+
+      if (names[len])
+         return false;
+   }
+
+   return true;
+}
+
 static void Settings_defaultMeters(Settings* this, unsigned int initialCpuCount) {
    int sizes[] = { 3, 3 };
 
@@ -313,7 +336,7 @@ static bool Settings_read(Settings* this, const char* fileName, unsigned int ini
       String_freeArray(option);
    }
    fclose(fd);
-   if (!didReadMeters) {
+   if (!didReadMeters || !Settings_validateMeters(this)) {
       Settings_defaultMeters(this, initialCpuCount);
    }
    return didReadAny;

